src/goto-symex/goto_state.h

```diff
@@ -24,6 +25,19 @@ Author: Romain Brenguier, romain.brenguier@diffblue.com
 // by the parent class.
 class goto_symex_statet;
 
+/// This is used for eliminating repeated complicated dereferences.
+/// This is pretty much just a simple wrapper around a map.
+/// \see goto_symext::dereference_rec
+struct dereference_cachet
+{
+private:
+  std::unordered_map<exprt, symbol_exprt, irep_hash> cache;
+
+public:
+  optionalt<symbol_exprt> lookup(const exprt &pointer) const;
+  void insert(exprt new_cached_pointer_expr, symbol_exprt new_cache_symbol);
+};
```

Comment:
Wouldn't you really want to use a shared map here?

Reply:
This is unclear and perhaps a premature optimisation at this point. For now we would like to merge this PR with minimal further experiments/updates that may have significant performance/behaviour impact.

Reply:
Why?

Reply:
Let me elaborate a bit: goto-symex is one of the most performance-critical pieces of code. It's not exactly a playground. Changes in here may impact every users of CBMC/JBMC. Changes to `goto_statet`, more specifically, have an effect upon every `GOTO` instruction simulated during symbolic execution. If data structures need to be copied, they need to be copied on every such occasion (and eventually be deleted again). With sharing maps, we have made considerable progress in making those copies cheap, both in terms of memory as well as CPU time. If we now introduce a non-shared data structure, even when it is empty, we will make copy and delete of `goto_statet` more expensive. Therefore, I have to ask: why does this need to be merged with minimal further experiments/updates?

Reply:
Implemented with sharing_map and pushed.

Reply:
Regarding the "Why": I don't have a clear answer as I did not make the initial decision. I was reluctant to modify since this could invalidate all the benchmarking, but a VERY preliminary run on a few examples did not show a problem. This does however come with the caveat that to some extent we may need to revisit SOME of the performance... but this is performance in the symex, which is typically negligible compared to SAT performance.

I hope this can satisfy the question, or at least answer it. I'll see if I can get some kind of performance data... but do not hold your breath over the weekend for this.

Reply:
Thanks for the further elaboration (your message only loaded after I posted mine). Good to better understand, and you're right in the sense that there is not a strong reason to merge immediately since this is performance critical. The sharing map is now used and we'll have some data on it next week.

Reply:
Data added below, seems to be a small improvement to use sharing_map. Also as per @martin-cs comments, I've used the sharing_map directly and removed the indirection.

Reply:
@tautschnig @martin-cs 
Updating this because there was a problem in the run data, it seems like changing from unordered_map to sharing_map has unstable impacts on runtime. See highlighted lines below.


Test | Develop | CSE | CSE_SM
-- | -- | -- | --
aws_add_size_checked | 5.85 | 5.63 | 5.72
aws_add_size_saturating | 5.74 | 5.71 | 5.70
aws_array_eq | 5.80 | 5.77 | 5.85
aws_array_eq_c_str | 5.82 | 5.72 | 5.76
aws_array_eq_c_str_ignore_case | 5.84 | 5.78 | 5.71
aws_array_eq_ignore_case | 5.78 | 5.88 | 5.80
aws_array_list_back | 20.21 | 14.08 | 18.21
aws_array_list_capacity | 11.98 | 12.02 | 10.00
aws_array_list_clean_up | 5.79 | 5.74 | 5.69
aws_array_list_clear | 5.71 | 5.79 | 5.57
aws_array_list_comparator_string | 7.65 | 7.82 | 7.95
aws_array_list_copy | 20.29 | 58.23 | 36.64
aws_array_list_ensure_capacity | 14.99 | 20.32 | 18.15
aws_array_list_erase | 334.31 | 346.68 | 323.94
aws_array_list_front | 9.94 | 10.03 | 7.86
aws_array_list_get_at | 22.24 | 11.99 | 33.65
aws_array_list_get_at_ptr | 17.21 | 14.06 | 15.08
aws_array_list_init_dynamic | 5.69 | 5.65 | 5.70
aws_array_list_init_static | 7.80 | 7.84 | 7.80
aws_array_list_length | 6.73 | 9.92 | 5.67
aws_array_list_pop_back | 13.89 | 14.06 | 26.27
**aws_array_list_pop_front** | 30.54 | 20.29 | 100.43
aws_array_list_pop_front_n | 26.41 | 24.37 | 24.31
**aws_array_list_push_back** | 86.01 | 55.18 | 47.94
aws_array_list_set_at | 40.84 | 40.91 | 32.52
aws_array_list_shrink_to_fit | 36.79 | 49.05 | 50.04
aws_array_list_sort | 33.57 | 32.57 | 22.15
aws_array_list_swap | 78.92 | 81.91 | 77.73
aws_array_list_swap_contents | 11.89 | 10.77 | 11.82
aws_byte_buf_advance | 7.80 | 7.78 | 7.82
aws_byte_buf_append | 5.80 | 5.71 | 5.78
**aws_byte_buf_append_dynamic** | 75.88 | 94.28 | 101.53
aws_byte_buf_append_with_lookup | 7.90 | 7.75 | 7.87
**aws_byte_buf_cat** | 76.82 | 56.20 | 79.93
aws_byte_buf_clean_up | 5.81 | 5.70 | 5.78
aws_byte_buf_clean_up_secure | 7.84 | 7.81 | 7.87
aws_byte_buf_eq | 6.71 | 5.72 | 5.68
aws_byte_buf_eq_c_str | 5.82 | 5.82 | 5.79
aws_byte_buf_eq_c_str_ignore_case | 5.87 | 5.88 | 5.74
aws_byte_buf_eq_ignore_case | 5.86 | 5.77 | 5.76
aws_byte_buf_from_array | 5.76 | 5.76 | 5.76
aws_byte_buf_from_c_str | 5.79 | 5.77 | 5.75
aws_byte_buf_from_empty_array | 5.76 | 5.73 | 5.75
aws_byte_buf_init | 5.73 | 5.76 | 5.79
aws_byte_buf_init_copy | 6.88 | 7.81 | 7.72
aws_byte_buf_init_copy_from_cursor | 5.77 | 5.75 | 5.76
aws_byte_buf_reserve | 5.77 | 5.81 | 5.66
aws_byte_buf_reserve_relative | 5.79 | 5.61 | 5.87
aws_byte_buf_reset | 5.85 | 5.71 | 5.74
aws_byte_buf_secure_zero | 9.86 | 9.87 | 10.03
aws_byte_buf_write | 11.98 | 13.98 | 11.98
aws_byte_buf_write_be16 | 5.79 | 5.71 | 5.78
aws_byte_buf_write_be32 | 5.64 | 5.85 | 5.75
aws_byte_buf_write_be64 | 5.77 | 5.89 | 5.80
aws_byte_buf_write_from_whole_buffer | 7.89 | 7.79 | 7.88
aws_byte_buf_write_from_whole_cursor | 5.85 | 5.72 | 5.83
aws_byte_buf_write_from_whole_string | 11.81 | 11.82 | 11.82
aws_byte_buf_write_u8 | 5.81 | 5.76 | 5.86
aws_byte_cursor_advance | 6.69 | 6.91 | 6.82
aws_byte_cursor_advance_nospec | 7.85 | 7.88 | 7.71
aws_byte_cursor_compare_lexical | 6.84 | 6.78 | 6.82
aws_byte_cursor_compare_lookup | 19.21 | 20.21 | 18.21
aws_byte_cursor_eq | 5.86 | 5.76 | 5.91
aws_byte_cursor_eq_byte_buf | 5.77 | 5.84 | 5.88
aws_byte_cursor_eq_byte_buf_ignore_case | 5.69 | 5.82 | 5.81
aws_byte_cursor_eq_c_str | 5.75 | 5.78 | 5.82
aws_byte_cursor_eq_c_str_ignore_case | 5.87 | 5.88 | 5.82
aws_byte_cursor_eq_ignore_case | 5.84 | 5.81 | 5.81
aws_byte_cursor_from_array | 5.89 | 5.80 | 5.79
aws_byte_cursor_from_buf | 5.83 | 5.85 | 5.80
aws_byte_cursor_from_c_str | 5.70 | 5.83 | 5.80
aws_byte_cursor_from_string | 5.70 | 5.63 | 5.69
aws_byte_cursor_left_trim_pred | 5.68 | 5.79 | 5.77
aws_byte_cursor_read | 11.99 | 9.82 | 9.82
aws_byte_cursor_read_and_fill_buffer | 14.05 | 14.09 | 13.79
aws_byte_cursor_read_be16 | 5.79 | 5.83 | 5.78
aws_byte_cursor_read_be32 | 5.78 | 5.76 | 5.79
aws_byte_cursor_read_be64 | 5.74 | 5.85 | 5.77
aws_byte_cursor_read_u8 | 5.79 | 5.85 | 5.87
aws_byte_cursor_right_trim_pred | 5.68 | 5.80 | 5.79
aws_byte_cursor_satisfies_pred | 5.82 | 5.80 | 5.70
aws_byte_cursor_trim_pred | 7.96 | 6.90 | 6.81
aws_hash_array_ignore_case | 5.88 | 5.70 | 5.72
aws_hash_byte_cursor_ptr | 5.86 | 5.76 | 5.78
aws_hash_byte_cursor_ptr_ignore_case | 5.79 | 5.68 | 5.87
aws_hash_c_string | 5.71 | 5.89 | 5.87
aws_hash_callback_c_str_eq | 7.91 | 7.84 | 7.91
aws_hash_callback_string_destroy | 5.67 | 5.74 | 5.75
aws_hash_callback_string_eq | 17.18 | 18.20 | 19.15
aws_hash_iter_begin | 9.95 | 9.97 | 14.09
aws_hash_iter_delete | 5.82 | 6.69 | 7.79
aws_hash_iter_done | 9.86 | 9.91 | 14.12
**aws_hash_iter_next** | 26.28 | 28.35 | 53.15
aws_hash_ptr | 5.82 | 5.74 | 5.76
aws_hash_string | 7.68 | 7.86 | 7.84
aws_hash_table_clean_up | 11.98 | 11.89 | 18.09
**aws_hash_table_clear** | 32.68 | 30.39 | 66.74
**aws_hash_table_create** | 86.35 | 86.07 | 227.89
**aws_hash_table_eq** | 40.86 | 36.73 | 115.93
**aws_hash_table_find** | 42.84 | 38.72 | 71.67
aws_hash_table_foreach | 9.91 | 7.84 | 9.80
aws_hash_table_get_entry_count | 5.76 | 5.78 | 5.73
aws_hash_table_init_bounded | 5.74 | 5.74 | 5.82
aws_hash_table_init_unbounded | 5.76 | 5.77 | 5.71
aws_hash_table_move | 5.82 | 5.82 | 5.74
**aws_hash_table_put** | 166.76 | 165.60 | 364.88
aws_hash_table_remove | 34.73 | 34.69 | 57.40
aws_hash_table_swap | 7.77 | 7.85 | 7.85
aws_is_power_of_two | 3.54 | 3.62 | 3.65
aws_linked_list_back | 9.95 | 9.85 | 9.80
aws_linked_list_begin | 9.86 | 9.71 | 11.85
aws_linked_list_end | 9.91 | 12.07 | 11.77
aws_linked_list_front | 9.88 | 9.93 | 9.69
aws_linked_list_init | 5.76 | 5.69 | 5.76
aws_linked_list_insert_after | 5.82 | 5.75 | 5.70
aws_linked_list_insert_before | 5.82 | 5.66 | 5.72
aws_linked_list_next | 5.71 | 5.83 | 5.85
aws_linked_list_node_reset | 5.89 | 5.78 | 5.75
**aws_linked_list_pop_back** | 896.49 | 927.29 | 22.19
aws_linked_list_pop_front | 16.05 | 16.05 | 18.13
aws_linked_list_prev | 5.69 | 5.62 | 5.75
aws_linked_list_push_back | 9.96 | 10.00 | 11.92
aws_linked_list_push_front | 14.01 | 12.83 | 14.04
aws_linked_list_rbegin | 9.88 | 9.91 | 12.02
aws_linked_list_remove | 5.82 | 5.82 | 5.71
aws_linked_list_rend | 9.89 | 9.97 | 11.96
aws_linked_list_swap_contents | 22.22 | 24.36 | 22.28
aws_mul_size_checked | 5.80 | 5.81 | 5.64
aws_mul_size_saturating | 5.80 | 5.64 | 5.72
aws_nospec_mask | 5.77 | 5.68 | 5.69
aws_priority_queue_capacity | 11.93 | 11.80 | 7.81
aws_priority_queue_clean_up | 5.73 | 5.77 | 5.71
aws_priority_queue_init_dynamic | 5.87 | 5.84 | 5.62
aws_priority_queue_init_static | 7.86 | 5.74 | 5.81
aws_priority_queue_pop | 61.45 | 51.17 | 51.17
**aws_priority_queue_push** | 688.75 | 634.17 | 5,713.86
**aws_priority_queue_push_ref** | 745.58 | 678.43 | 770.18
**aws_priority_queue_remove** | 60.40 | 123.28 | 71.76
**aws_priority_queue_s_remove_node** | 1,945.39 | 2,157.15 | 1,770.73
**aws_priority_queue_s_sift_down** | 397.25 | 415.76 | 450.51
**aws_priority_queue_s_sift_either** | 242.71 | 325.22 | 246.72
aws_priority_queue_s_sift_up | 40.61 | 36.73 | 36.64
aws_priority_queue_s_swap | 395.18 | 415.94 | 408.40
aws_priority_queue_size | 7.92 | 9.90 | 9.91
aws_priority_queue_top | 30.48 | 16.04 | 16.09
aws_ptr_eq | 3.74 | 3.64 | 4.66
aws_ring_buffer_acquire | 60.54 | 58.24 | 56.24
aws_ring_buffer_acquire_up_to | 72.83 | 72.81 | 72.66
aws_ring_buffer_buf_belongs_to_pool | 18.50 | 18.66 | 18.51
aws_ring_buffer_clean_up | 5.82 | 5.85 | 5.89
aws_ring_buffer_init | 5.71 | 5.81 | 5.75
aws_ring_buffer_release | 29.04 | 27.07 | 27.05
aws_round_up_to_power_of_two | 3.71 | 3.61 | 3.72
aws_string_bytes | 5.74 | 5.71 | 5.71
aws_string_compare | 7.78 | 7.85 | 6.79
aws_string_destroy | 5.68 | 5.75 | 5.67
aws_string_destroy_secure | 5.72 | 5.74 | 5.68
aws_string_eq | 7.85 | 7.87 | 5.76
aws_string_eq_byte_buf | 7.80 | 7.70 | 7.87
aws_string_eq_byte_buf_ignore_case | 6.71 | 6.72 | 6.80
aws_string_eq_byte_cursor | 7.80 | 7.64 | 7.87
aws_string_eq_byte_cursor_ignore_case | 7.78 | 6.76 | 7.84
aws_string_eq_c_str | 5.89 | 5.92 | 5.80
aws_string_eq_c_str_ignore_case | 5.80 | 5.83 | 5.63
aws_string_eq_ignore_case | 7.92 | 7.85 | 6.85
aws_string_new_from_array | 7.87 | 7.86 | 7.80
aws_string_new_from_c_str | 5.89 | 5.74 | 5.70
aws_string_new_from_string | 11.89 | 11.77 | 9.79
memcpy_using_uint64 | 27.35 | 28.42 | 28.28
memset_override_0 | 25.36 | 26.27 | 26.23
memset_using_uint64 | 35.63 | 36.66 | 34.52
**Total** | 8,092.01 | 8,365.22 | 12,756.92

This makes it unclear if using sharing_map is overall a significant advantage or not.

I'm gathering more data for this and other experiments, but the evidence here is unclear on whether sharing_map is a "better" choice than unordered_map.